Why do we use Relay? (#553)
artsy/artsy.github.io: _posts/2019-04-06-omakase-relay.markdown

---
layout: epic
title: "Why does Artsy use Relay?"
date: "2019-04-06"
author: [orta]
categories: [community, omakase, relay]
---

> When the mobile team at Artsy considered moving to React Native back in 2014, one of the most compelling cases for

Reviewer: `React Native back in 2015` -- wasn't announced yet in 2014

making that jump was Relay. This, it seems, is a dependency that is rarely used in the JS community and we often
find ourselves defending this decision to new engineers during onboarding. It's great to [..]

Let's have a deep dive into what makes Relay compelling for Artsy Engineering.

<!-- more -->

# What problem does Relay solve?

Relay is an API client for GraphQL, it comes in two parts: a compiler and a set of front-end components. Relay aims
to provide a really tight binding between your GraphQL API and your view hierarchy. When you build data-driven
apps, Relay removes a whole suite of non-business logic from your application.

Relay handles:

- Data binding (API -> props)

> - Data transformations (response shaping)

Reviewer: What do you mean by this exactly?

Author: Good point. I think it doesn't add enough value to be in there. But my original intention was to encompass local schema extensions (without the direct jargon) but we don't use it and it's an arcane API ( https://github.com/facebook/relay/issues/1787 ) so I'm killing it

Facebook-scale best-practices and can build on top of that.

# How does it work?

You write a set of Relay components, you always start with a `QueryRenderer` then use a tree of either
`FragmentContainer`, `RefetchContainer` or `PaginationController`s. You mostly use `FragmentContainer`s.

The `FragmentContainer` is based on a GraphQL fragment. If you've never used a fragment, they are an abstraction
that lets you declare shared fields on a specific GraphQL type to reduce duplication up your queries. For example:

```
query GetPopularArtistsAndMyFavs {
  me {
    artists {
      id
      name
      bio
    }
  }
  popularArtists {
    id
    name
    bio
  }
}
```

To move this query to use fragments:

```
query GetPopularArtistsAndMyFavs {
  me {
    artists {
      ...ArtistMetadata
    }
  }
  popularArtists {
    ...ArtistMetadata
  }
}

fragment ArtistMetadata on Artist {
  id
  name
  bio
}
```

It's tiny a bit longer, but you have a guarantee that the data is consistent across both sets of artists. Now that
you have a rough glimpse at what a GraphQL fragment is, let's look at what a `FragmentContainer` looks like. Here's

> a simplified [profile page][] from the Artsy iOS app:

Reviewer: Additional `[]`

```ts
import React from "react"
import { createFragmentContainer, graphql } from "react-relay"
import { MyProfile_me } from "__generated__/MyProfile_me.graphql"

> // [...]

Reviewer: Prob not necessary to communicate the point

    )
  }
}

export default createFragmentContainer(
  MyProfile,
  graphql`
    fragment MyProfile_me on Me {
      name
      image
      initials
    }
  `
)
```

There are three moving parts:

- The `MyProfile` component, which is a vanilla React component
- The exported `createFragmentContainer` which wraps the `MyProfile` and ties it to a fragment on a `Me` type in
  GraphQL.
- The TypeScript interface `MyProfile_me` which ensures we have a correct interface to our props

## Isolation

The React component `MyProfile` will be passed in props that directly tie to the fragment that was requested. In
Relay terms, this is called data-masking and it is one of the first hurdles for someone new to Relay to [grok][].
In REST clients, and GraphQL API clients like Apollo Client, you make a request and that request is passed through
the React tree. E.g.

{% include epic_img.html url="/images/omakase-relay/tree.png" title="REST inspired props" style="width:100%;" %}

This means most components know more about the request than it probably needs, as it may be needed to pass on to
the component's children. This can lead to data-duplication, or even worse, not knowing if you can delete or

> refactor a component.

Reviewer: Ooh very good point 👍

Data masking solves this by hiding data that the component didn't request. I've still yet to find the right visual
abstraction, but I feel this just about pays for itself.

{% include epic_img.html url="/images/omakase-relay/isolation.png" title="Relay isolation tree" style="width:100%;" %}

You let Relay be responsible for consolidating all your fragments into a query, request it, and (mostly) passing it
through your component hierarchy. This means Relay powered component can be safely changed and drastically reduces
the chance for unintended consequences elsewhere.

This isolation gives Artsy engineers the safety to work on projects with tens of contributors over long time

> periods without too much technical debt. The components we create are nearly all focused only on the data-driven

Reviewer: I wish there was a better / more elaborated way to express just how sturdy, and just how little technical debt (at least at the data layer) there typically is.

<img src="/images/omakase-relay/co-location.png">

In roughly that proportion too, though our most modern code uses the Artsy design system [Palette][palette] which
drastically reduces the need for styling.

## Community

When we adopted Relay, there was no competition - we'd have just used the fetch API. Over time, Apollo came up and
really put a considerable amount of effort into lowering the barriers to entry, and making it feasible to build
complex apps easily.

We did an audit last year of what it would take to re-create a lot of the infrastructure we use in Relay atop of
the (much more popular) Apollo GraphQL eco-system and saw it was feasible but would require a considerable amount
of work across many different plugins and tools. With Relay that's all packaged into one tool, works consistently
and obviously doesn't have a scaling problem as Facebook have tens of thousands of Relay components.

## Scale Safety

Relay puts a lot of emphasis on ahead of time safety. The Relay compiler validates your queries against your
GraphQL schema, we extended the compiler to create TypeScript types for the composed API fragments and there are

> strict naming systems enforced by the compiler. All of these guides engineers to build scalable codebases.

Reviewer: I'm sure most people understand what "letting the compiler drive" means but maybe it would be cool to add a little screen cap showing how invalid GraphQL code fails at the compiler level and not at runtime? Apollo has similar validation features, but not without some setup.

Author: Done!

---
layout: epic
title: "Why does Artsy use Relay?"
date: "2019-04-06"
author: [orta]
categories: [community, omakase, relay]
css: relay
comment_id: 557
---

When the mobile team at Artsy considered moving to React Native back in 2016, one of the most compelling cases for
making that jump was Relay. This, it seems, is a dependency that is rarely used in the JS community and we often
find ourselves defending this decision to new engineers during onboarding and to the public at large.

Which makes this a perfect blog post topic, so let's have a deep dive into what makes Relay compelling for Artsy's
engineering team.

<!-- more -->

# What problem does Relay solve?

Relay is an API client for GraphQL, it comes in two parts: a compiler and a set of front-end components. Relay aims
to provide a really tight binding between your GraphQL API and your view hierarchy. When you build data-driven
apps, Relay removes a whole suite of non-business logic from your application.

Relay handles:

- Data binding (API -> props)

> - Cache management (invalidation, updates etc)

Author: I am, yep

> - Cache management (invalidation, updates etc)

Reviewer 2: Perhaps this better explains that?

```suggestion
- Cache management (invalidation, re-render live components with data updates, etc)
```

- Consistent bi-directional pagination abstractions
- Multiple query consolidation (e.g. consolidate all API requests to one request)

> - Declarative data mutation (describe how data should change, instead of doing it)

Author: we chatted about this IRL, and I asked in slack - https://graphql.slack.com/archives/C0BEXJLKG/p1554740176044300

Removing this line, as it's moving out of core

> - Declarative data mutation (describe how data should change, instead of doing it)

Reviewer 2: Didn’t sound like they are going to be considered deprecated, though. Perhaps we _should_ use them?

> - UI best practices baked in (optimistic response rendering, cheap rollbacks)

Author: Mainly the diff between the optimistic response and the real response, but it's a React feature - not relays, removing

- AOT query generation (allowing you to persist queries)

By taking the responsibilities of the grunt work for most complex apps and moving it into Relay you get

> Facebook-scale best-practices and can build on top of that.

Author: exactly

# How does it work?

You write a set of Relay components, you always start with a [`QueryRenderer`][query] then use a tree of either
[`FragmentContainer`][frag], [`RefetchContainer`][re] or [`PaginationController`][pag]s. You mostly use
`FragmentContainer`s so I'll focus on that here.

A `FragmentContainer` is based on a [GraphQL fragment][gql-frag]. If you've never used a fragment, they are an
abstraction that lets you declare shared fields on a specific GraphQL type to reduce duplication up your queries.
For example:

```
query GetPopularArtistsAndMyFavs {
  me {

>     artists {

Author: moved this to be two singular fields, for simplicity

      id
      name
      bio
    }
  }
  popularArtists {
    id
    name
    bio
  }
}
```

To move this query to use fragments:

```
query GetPopularArtistsAndMyFavs {
  me {
    artists {
      ...ArtistMetadata
    }
  }
  popularArtists {
    ...ArtistMetadata
  }
}

fragment ArtistMetadata on Artist {
  id
  name
  bio
}
```

It's tiny a bit longer, but you have a guarantee that the data is consistent across both sets of artists. Now that
you have a rough idea of what a GraphQL fragment is, let's look at what a `FragmentContainer` looks like. Here's a
simplified [profile page] from the Artsy iOS app:

```ts
import React from "react"
import { createFragmentContainer, graphql } from "react-relay"
import { MyProfile_me } from "__generated__/MyProfile_me.graphql"

interface Props extends ViewProperties {
  me: MyProfile_me
}

export class MyProfile extends React.Component<Props> {
  render() {
    return (
      <View>
        <Header>

>           <ProfilePhoto initials={props.me.initials} image={props.me.image} />

Author: It could, but it is meant to be a simplification of the existing Component, so differing that much (by moving to a func ) probably would detract from the point

            onPress={goToUserSettings}
          />
        </ButtonSection>
      </View>
    )
  }
}

export default createFragmentContainer(
  MyProfile,
  graphql`
    fragment MyProfile_me on Me {
      name
      image
      initials
    }
  `
)
```

There are three moving parts:

- The TypeScript interface `MyProfile_me` which ensures we have a correct interface to our props
- The `MyProfile` component, which is a vanilla React component
- The exported `createFragmentContainer` which wraps the `MyProfile` and ties it to a fragment on a `Me` type in
  GraphQL.

## Isolation

The React component `MyProfile` will be passed in props that directly tie to the fragment that was requested. In
Relay terms, this is called data-masking and it is one of the first hurdles for someone new to Relay to [grok][].
In REST clients, and GraphQL API clients like Apollo Client, you make a request and that request is passed through
the React tree. E.g.

{% include epic_img.html url="/images/omakase-relay/tree.png" title="REST inspired props" style="width:100%;" %}

This means most components know more about the request than it probably needs, as it may be needed to pass on to

> the component's children. This can lead to data-duplication, or even worse, not knowing if you can delete or

Author: yeah, I was, good point

really put a considerable amount of effort into lowering the barriers to entry, and making it feasible to build
complex apps easily.

We did an audit last year of what it would take to re-create a lot of the infrastructure we use in Relay atop of
the (much more popular) Apollo GraphQL eco-system and saw it was feasible but would require a considerable amount
of work across many different plugins and tools. With Relay that's all packaged into one tool, works consistently
and obviously doesn't have a scaling problem as Facebook have tens of thousands of Relay components.

It's worth highlighting the core difference in community management for Apollo vs Relay. Engineers working on
Apollo have great incentives to do user support, and improve the tools for the community - that's their businesses
value. Relay on the other hand is used in many places at Facebook, and the engineers on the team support internal
issues first. IMO, this is reasonable, Relay is an opinionated batteries-included framework for building user
interfaces, and ensuring it works with the baffling amount of JavaScript at Facebook is more or less all the team
has time for.

> That leaves space for the OSS community to own their own problems.

Reviewer: Might be worth mentioning the `relay-tools` repo here.

## Scale Safety

Relay puts a lot of emphasis on ahead of time safety. The Relay compiler validates your queries against your
GraphQL schema, we extended the compiler to create TypeScript types for the composed API fragments and there are
strict naming systems enforced by the compiler. All of these guides engineers to build scalable codebases.

How this works in practice is that whenever you need to change the data a component requires, you edit the
fragment, the Relay compiler verifies your query, if successful then your TypeScript types are updated and you can
use the new property in your React component above. See below for a [quick video][vid] showing the Relay compiler
in action:

{% include epic_video.html url="/images/omakase-relay/relay-process-720.mov" title="Relay isolation tree" style="width:100%; display:block;" %}

## Cultural Fit

Relay fit well into our team because:

- We had engineers who were interested in contributing back and making it work for our cases, this is much less of

>   an issue now that Relay has matured and is better documented.

Reviewer 2: Not sure I understand how this relates to it being a cultural fit.

> - We had engineers used to using ahead-of-time error validation tools like compilers

Reviewer 2: ```suggestion
- We had engineers that were used to the benefits of ahead-of-time error validation tools like compilers.
```

> - We saw a lot of value in a tightly coupling our view structure to our user interface

Reviewer 2: ```suggestion
- We saw a lot of value in a tightly coupling our view structure to our user interface.
```

Not sure I really understand what this is trying to convey. View structure on disk?

> It's not without its flaws, but Relay has definitely paid for it's initial and occasional complexity for the

Reviewer 2: ```suggestion
It's not without its shortcomings [to users outside of Facebook], but Relay has definitely paid for its initial and occasional complexity for the
```

queries. For example:

```
query GetPopularArtistAndFeaturedArtist {
  featuredArtist {
    id
    name
    bio
  }
  popularArtist {
    id
    name
    bio
  }
}
```

To move this query to use fragments:

```
query GetPopularArtistsAndMyFavs {
  featuredArtist {
    ...ArtistMetadata
  }
  popularArtist {
    ...ArtistMetadata
  }
}

fragment ArtistMetadata on Artist {
  id
  name
  bio
}
```

It's a tiny bit longer, but you have a guarantee that the data is consistent across both sets of artists. Now that
you have a rough idea of what a GraphQL fragment is, let's look at what a `FragmentContainer` looks like. Here's a
simplified [profile page] from the Artsy iOS app:

```ts
import React from "react"
import { createFragmentContainer, graphql } from "react-relay"
import { MyProfile_me } from "__generated__/MyProfile_me.graphql"

interface Props extends ViewProperties {
  me: MyProfile_me
}

export class MyProfile extends React.Component<Props> {
  render() {
    return (
      <View>
        <Header>
          <ProfilePhoto initials={this.props.me.initials} image={this.props.me.image} />

>           <Subheading>{props.me.name}</Subheading>

Reviewer 2: ```suggestion
          <Subheading>{this.props.me.name}</Subheading>
```

        </Header>
        <ButtonSection>
          <ProfileButton
            section="Selling"
            description="Sell works from your collection"
            onPress={startSubmission}
          />
          <ProfileButton
            section="Account Details"
            description="Email, password reset, profile"
            onPress={goToUserSettings}
          />
        </ButtonSection>
      </View>
    )
  }
}

export default createFragmentContainer(
  MyProfile,
  graphql`
    fragment MyProfile_me on Me {
      name
      image
      initials
    }
  `
)
```

There are three moving parts:

- The TypeScript interface `MyProfile_me` which ensures we can only use fields that were selected in the fragment
- The `MyProfile` component, which is a vanilla React component
- The exported `createFragmentContainer` which returns a higher-order component that wraps `MyProfile` and ties it

>   to a fragment on a `Me` type in GraphQL.

Reviewer 2: If I didn’t misread, all other lists use no periods.

```suggestion
  to a fragment on a `Me` type in GraphQL
```

the styles, the actual view content hierarchy, and the exact parts of the API it needs to render itself.

<img src="/images/omakase-relay/co-location.png">

In roughly that proportion too, though our most modern code uses the Artsy design system [Palette][palette] which
drastically reduces the need for style in our components.

Co-location's biggest selling point is reducing [cognitive load](https://en.wikipedia.org/wiki/Cognitive_load),
having everything you need in one place makes it easier to understand how a component works. This makes code review
simpler, and lowers the barrier to understanding the entire systems at scale.

## Community

When we adopted Relay, there was no competition - we'd have just used the `fetch` API. Over time, [the Apollo
team][apollo] came up and really put a considerable amount of effort into lowering the barriers to entry, and
making it feasible to build complex apps easily.

As we hired a set of new engineers, the "Apollo vs Relay" debate came up. Interested in whether we would still
start with Relay today, we ran an audit last year of what it would take to re-create a lot of the infrastructure we

> love in Relay atop of the (much more popular) Apollo GraphQL eco-system and saw it was feasible but would require a

Reviewer 2: At its core, Apollo remains a lib that does more at runtime vs Relay. I don’t think we need to go into that detail, but I also wouldn’t say we could get 100% the same thing if we built enough atop of Apollo.

```suggestion
love in Relay atop of the (much more popular) Apollo GraphQL eco-system and saw it was reasonably feasible but would require a
```

considerable amount of work across many different plugins and tools. With Relay that's all packaged into one tool,

> works consistently and has been proven in production with Facebook have tens of thousands of Relay components.

Reviewer 2: ```suggestion
works consistently and has been proven with Facebook having tens of thousands of Relay components in production.
```
